torch/onnx/verify.py

```diff
@@ -0,0 +1,262 @@
+    if training is not None and training == torch.onnx.TrainingMode.TRAINING:
+        model.train()
+    elif training is None or training == torch.onnx.TrainingMode.EVAL:
+        model.eval()
+    if input is None:
+        input = torch.randn(batch_size, 3, 224, 224, requires_grad=True)
+    with torch.no_grad():
+        if isinstance(input, (Tensor, dict)):
+            input = (input,)
+        # In-place operators will update input tensor data as well.
+        # Thus inputs are replicated before every forward call.
+        input_args = copy.deepcopy(input)
+        input_kwargs = {}
+        if dict_check and isinstance(input_args[-1], dict):
+            input_kwargs = input_args[-1]
+            input_args = input_args[:-1]
+        try:
+            model_copy = copy.deepcopy(model)
+            output = model_copy(*input_args, **input_kwargs)
+        except Exception:
+            output = model(*input_args, **input_kwargs)
+        if isinstance(output, Tensor):
+            output = (output,)
+
+        if not dict_check and isinstance(input[-1], dict):
+            input = input + ({},)
+
+        with ExitStack() as stack:
+            model_f: Union[str, io.BytesIO] = io.BytesIO()
+            if use_external_data:
+                tmpdirname = stack.enter_context(tempfile.TemporaryDirectory())
+                model_f = os.path.join(tmpdirname, "model.onnx")
+
+            ort_sess = _convert_to_onnx(
+                model,
+                model_f,
+                input=input,
+                opset_version=opset_version,
+                do_constant_folding=do_constant_folding,
+                keep_initializers_as_inputs=keep_initializers_as_inputs,
+                dynamic_axes=dynamic_axes,
+                input_names=input_names,
+                output_names=output_names,
+                fixed_batch_size=fixed_batch_size,
+                training=training,
+                verbose=verbose,
+                ort_providers=ort_providers,
+                ort_optim_on=ort_optim_on,
+            )
+            # compute onnxruntime output prediction
+            if remained_onnx_input_idx is not None:
+                input_onnx = []
+                for idx in remained_onnx_input_idx:
+                    input_onnx.append(input[idx])
+                input = input_onnx
+
+            input_copy = copy.deepcopy(input)
+            if flatten:
+                input_copy, _ = torch.jit._flatten(input_copy)
+            elif input_copy and input_copy[-1] == {}:
+                # Handle empty kwargs (normally removed by flatten).
+                input_copy = input_copy[:-1]
+            ort_outs = _run_ort(ort_sess, input_copy)
+            _ort_compare_with_pytorch(ort_outs, output, rtol, atol)
+
+            # if additional test inputs are provided run the onnx
+            # model with these inputs and check the outputs
+            if test_with_inputs is not None:
```

Comment:
This function is a little too big. Would it be possible to factor some parts out? For example, the part that prepares the input (247-260) seems to be a good candidate.